fgpyo/vcf/__init__.py

```diff
@@ -0,0 +1,464 @@
+        return sd
+
+    @classmethod
+    def build_header_string(
+        cls, samples: Optional[List[str]] = None, sd: Optional[Dict[str, Dict[str, Any]]] = None
+    ) -> str:
+        """Builds the VCF header with the given sample name(s) and sequence dictionary.
+
+        Args:
+            samples: the sample name(s).
+            sd: the sequence dictionary mapping the contig name to the key-value pairs for the
+                given contig.  Must include "ID" and "length" for each contig.  If no sequence
+                dictionary is given, will use the default dictionary.
+        """
+        if sd is None:
+            sd = VariantBuilder.default_sd()
+        buffer = "##fileformat=VCFv4.2\n"
+        buffer += '##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">\n'
+        for d in sd.values():
+            assert "ID" in d, "ID missing in sequence dictionary"
+            assert "length" in d, "length missing in sequence dictionary"
+            contig_id = d["ID"]
+            contig_length = d["length"]
+            contig_buffer = f"##contig=<ID={contig_id},length={contig_length}"
+            for key, value in d.items():
+                if key == "ID" or key == "length":
+                    continue
+                contig_buffer += f",{key}={value}"
+            contig_buffer += ">\n"
+            buffer += contig_buffer
+
+        buffer += "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT"
+        if samples is not None:
+            buffer += "\t"
+            buffer += "\t".join(samples)
+        buffer += "\n"
+
+        return buffer
+
+    def __init__(
+        self, samples: Optional[List[str]] = None, sd: Optional[Dict[str, Dict[str, Any]]] = None
+    ) -> None:
+        """Initializes a new VariantBuilder for generating variants and VCF files.
+
+        Args:
+            samples: the name of the sample(s)
+            sd: optional sequence dictionary
+        """
+        self.samples: List[str] = samples if samples is not None else []
+        self.sd: Dict[str, Dict[str, Any]] = sd if sd is not None else VariantBuilder.default_sd()
+        self.seq_idx_lookup: Dict[str, int] = {name: i for i, name in enumerate(self.sd.keys())}
+        self.records: List[Variant] = []
+
+    def __enter__(self) -> "VariantBuilder":
+        return self
+
+    def __exit__(self, *args: Any) -> bool:  # type: ignore
+        self.close()
+        return False
+
+    @abstractmethod
+    def close(self) -> None:
+        """Closes this builder"""
+        pass
+
+    @abstractmethod
+    def add(
+        self,
+        contig: Optional[str] = None,
+        pos: Optional[int] = None,
+        id: Optional[str] = None,
+        ref: Optional[str] = None,
+        alts: Optional[Iterable[str]] = None,
+        qual: Optional[int] = None,
+        filter: Optional[Iterable[str]] = None,
+        info: Optional[Dict[str, Any]] = None,
+        samples: Optional[Dict[str, Dict[str, Any]]] = None,
```

Comment:
I updated the docstring to be more explicit. You can omit fields (they will be set to no-call). You can omit samples (all fields will be set to no-call). I think this way is nicer, because otherwise it's tough to have builder return a `pysam.VariantRecord`.